File: homeworks/Nigmatullin/stack/delphi unit/stack.pas
unit stack;

interface

const
    N = 100;

type
    index = 0..N;
    T = integer;

    myStack = record
        data : array[1..N] of T;
        current : index;
    end;
    
procedure push(var s : myStack; el : T);
function isEmpty(var s : myStack) : boolean;
function pop(var s : myStack) : T;
procedure empty(var s : myStack);



implementation

procedure push(var s : myStack; el : T);
begin
    if s.current < N then
    begin
        Inc(s.current);
        s.data[s.current] := el;
    end;
end;


function isEmpty(var s : myStack) : boolean;
begin
    isEmpty := not (s.current > 0);
end;

function pop(var s : myStack) : T;
begin
    if not isEmpty(s) then
    begin
        pop := s.data[s.current];
        Dec(s.current);
    end
    else
    begin
        pop := -1;
    end;
end;

procedure empty(var s : myStack);
begin
    s.current := 0;
end;

end.

File: homeworks/Nigmatullin/stack/cpp lib/Stack.cpp
#include "Stack.h"

stack::stack()
{
	current = -1;
}

void push(stack &s, int el)
{
	if (s.current < N - 1)
	{
		s.data[++s.current] = el;
	}
}

bool isEmpty(stack &s)
{
	return !(s.current > -1);
}

int pop(stack &s)
{
	if (!isEmpty(s))
	{
		return s.data[s.current--];
	}
	else
	{
		return -1;
	}
}

void empty(stack &s)
{
	s.current = -1;
}

